pkg/proxy/proxy.go

```diff
@@ -0,0 +1,218 @@
+}
+
+func (p *Proxy) handleRequestAndRedirect(res http.ResponseWriter, req *http.Request) {
+	ctx, err := p.createContext(req)
+	if err != nil {
+		log.Error(nil, err, "unable to create a context")
+		responseWithError(res, crterrors.NewUnauthorizedError("unable to create a context", err.Error()))
+		return
+	}
+	ns, err := p.getTargetNamespace(ctx)
+	if err != nil {
+		log.Error(ctx, err, "unable to get target namespace")
+		responseWithError(res, crterrors.NewInternalError(errors.New("unable to get target namespace"), err.Error()))
+		return
+	}
+
+	p.serveReverseProxy(ctx, ns, res, req)
+}
+
+func responseWithError(res http.ResponseWriter, err *crterrors.Error) {
+	http.Error(res, err.Error(), err.Code)
+}
+
+// createContext creates a new gin.Context with the User ID extracted from the Bearer token.
+// To be used for storing the user ID and logging only.
+func (p *Proxy) createContext(req *http.Request) (*gin.Context, error) {
+	userID, err := p.extractUserID(req)
+	if err != nil {
+		return nil, err
+	}
+	keys := make(map[string]interface{})
+	keys[context.SubKey] = userID
+	return &gin.Context{
+		Keys: keys,
+	}, nil
+}
+
+func (p *Proxy) getTargetNamespace(ctx *gin.Context) (*namespace.Namespace, error) {
+	userID := ctx.GetString(context.SubKey)
+	return p.namespaces.GetNamespace(ctx, userID)
+}
+
+func (p *Proxy) extractUserID(req *http.Request) (string, error) {
+	userToken, err := extractUserToken(req)
+	if err != nil {
+		return "", err
+	}
+
+	token, err := p.tokenParser.FromString(userToken)
+	if err != nil {
+		return "", crterrors.NewUnauthorizedError("unable to extract userID from token", err.Error())
+	}
+	return token.Subject, nil
+}
+
+func extractUserToken(req *http.Request) (string, error) {
+	a := req.Header.Get("Authorization")
+	token := strings.Split(a, "Bearer ")
+	if len(token) < 2 {
+		return "", crterrors.NewUnauthorizedError("no token found", "a Bearer token is expected")
+	}
+	return token[1], nil
+}
+
+func (p *Proxy) serveReverseProxy(ctx *gin.Context, target *namespace.Namespace, res http.ResponseWriter, req *http.Request) {
+	proxy := p.newReverseProxy(ctx, target)
+
+	// Note that ServeHttp is non blocking and uses a go routine under the hood
+	proxy.ServeHTTP(res, req)
+}
+
+func (p *Proxy) newReverseProxy(ctx *gin.Context, target *namespace.Namespace) *httputil.ReverseProxy {
+	targetQuery := target.APIURL.RawQuery
+	director := func(req *http.Request) {
+		origin := req.URL.String()
+		req.URL.Scheme = target.APIURL.Scheme
+		req.URL.Host = target.APIURL.Host
+		req.URL.Path = singleJoiningSlash(target.APIURL.Path, req.URL.Path)
+		log.Info(ctx, fmt.Sprintf("forwarding %s to %s", origin, req.URL.String()))
+		if targetQuery == "" || req.URL.RawQuery == "" {
+			req.URL.RawQuery = targetQuery + req.URL.RawQuery
+		} else {
+			req.URL.RawQuery = targetQuery + "&" + req.URL.RawQuery
+		}
+		if _, ok := req.Header["User-Agent"]; !ok {
+			// explicitly disable User-Agent so it's not set to default value
```

Comment:
why do we need to set `User-Agent` to empty value when it's missing?

Reply:
As the comment is saying its to avoid using the default value. But I don't know why we would need that. It's something I saw in the reverse proxy examples. Let's keep it here until we find out we don't want or need that :)